lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,239 @@
+  FixedDateTimeFormatter(this.pattern) {
+    int? current;
+    var start = 0;
+    var characters = pattern.codeUnits;
+    for (var i = 0; i < characters.length; i++) {
+      var char = characters[i];
+      if (current != char) {
+        _blocks.saveBlock(current, start, i);
+        if (_validFormatCharacters.contains(char)) {
+          var hasSeenBefore = _blocks.formatCharacters.indexOf(char);
+          if (hasSeenBefore > -1) {
+            throw FormatException(
+                "Pattern contains more than one '$char' block.\n"
+                "Previous occurrence at index ${_blocks.starts[hasSeenBefore]}",
+                pattern,
+                i);
+          } else {
+            start = i;
+            current = char;
+          }
+        } else {
+          current = null;
+        }
+      }
+    }
+    _blocks.saveBlock(current, start, pattern.length);
+  }
+
+  /// Convert [DateTime] to a [String] exactly as specified by the [pattern].
+  String encode(DateTime datetime) {
+    var buffer = StringBuffer();
+    for (var i = 0; i < _blocks.length; i++) {
+      var start = _blocks.starts[i];
+      var end = _blocks.ends[i];
+      var length = end - start;
+
+      var previousEnd = i > 0 ? _blocks.ends[i - 1] : 0;
+      if (previousEnd < start) {
+        buffer.write(pattern.substring(previousEnd, start));
+      }
+      var number = _extractNumFromDateTime(
+        _blocks.formatCharacters[i],
+        datetime,
+      ).toString();
+      if (number.length > length) {
+        number = number.substring(number.length - length);
+      } else if (number.length < length) {
+        number = number.padLeft(length, '0');
+      }
+      buffer.write(number);
+    }
+    if (_blocks.length > 0) {
+      var lastEnd = _blocks.ends.last;
+      if (lastEnd < pattern.length) {
+        buffer.write(pattern.substring(lastEnd, pattern.length));
+      }
+    }
+    return buffer.toString();
+  }
+
+  int _extractNumFromDateTime(int? formatChar, DateTime dt) {
+    if (formatChar == yearCode) {
+      return dt.year;
+    } else if (formatChar == centuryCode) {
+      return (dt.year / 100).floor();
+    } else if (formatChar == decadeCode) {
+      return (dt.year / 10).floor();
+    } else if (formatChar == monthCode) {
+      return dt.month;
+    } else if (formatChar == dayCode) {
+      return dt.day;
+    } else if (formatChar == hourCode) {
+      return dt.hour;
+    } else if (formatChar == minuteCode) {
+      return dt.minute;
+    } else if (formatChar == secondCode) {
+      return dt.second;
+    }
+    throw AssertionError("Unreachable, the key is checked in the constructor");
+  }
+
+  /// Parse a string [formattedDateTime] to a local [DateTime] as specified in the
+  /// [pattern], substituting missing values with a default. Throws an exception
+  /// on failure to parse.
+  DateTime decode(String formattedDateTime, {bool isUtc = false}) {
```

Comment:
Should we default `isUtc` to true instead? It's the better choice for representing a textually-given date in most cases.